server/app/views/components/FieldWithLabel.java

```diff
@@ -246,107 +243,183 @@ public FieldWithLabel showFieldErrors(boolean showFieldErrors) {
+    if (fieldTag.getTagName().equals("textarea")) {
+      fieldTag.withText(this.fieldValue);
+    }
+
+    return applyAttrsAndGenLabel(fieldTag);
+  }
+
+  protected DivTag getTagContainer(Tag fieldTag) {
+    genRandIdIfEmpty();
+    if (this.fieldType.equals("number")) {
+      numberTagApplyAttrs(fieldTag);
+    } else {
+      fieldTag.attr("value", this.fieldValue);
+    }
+
+    return applyAttrsAndGenLabel(fieldTag);
+  }
+
+  private DivTag buildFieldErrorsTag(String id) {
+    String[] referenceClasses =
+        referenceClassesBuilder.build().stream().map(ref -> ref + "-error").toArray(String[]::new);
+    return div(each(fieldErrors, error -> div(error.format(messages))))
+        .withId(id)
+        .withClasses(
+            StyleUtils.joinStyles(referenceClasses),
+            StyleUtils.joinStyles(BaseStyles.FORM_ERROR_TEXT_XS, Styles.P_1),
+            fieldErrors.isEmpty() || !showFieldErrors ? Styles.HIDDEN : "");
+  }
+
+  protected void genRandIdIfEmpty() {
     // In order for the labels to be associated with the fields (mandatory for screen readers)
     // we need an id.  Generate a reasonable one if none is provided.
     if (this.id.isEmpty()) {
       this.id = RandomStringUtils.randomAlphabetic(8);
     }
-    String fieldErrorsId = String.format("%s-errors", this.id);
-    if (fieldTag.getTagName().equals("textarea")) {
-      // Have to recreate the field here in case the value is modified.
-      ContainerTag textAreaTag = textarea().withType("text").withText(this.fieldValue);
-      fieldTag = textAreaTag;
-    } else if (this.fieldType.equals("number")) {
-      // Setting inputmode to decimal gives iOS users a more accessible keyboard
-      fieldTag.attr("inputmode", "decimal");
-
-      // Setting step to any disables the built-in HTML validation so we can use our
-      // custom validation message to enforce integers.
-      fieldTag.attr("step", "any");
-
-      // Set min and max values for client-side validation
-      if (this.minValue.isPresent()) {
-        fieldTag.attr("min", minValue.getAsLong());
-      }
-      if (this.maxValue.isPresent()) {
-        fieldTag.attr("max", maxValue.getAsLong());
-      }
-
-      // For number types, only set the value if it's present since there is no empty string
-      // equivalent for numbers.
-      if (this.fieldValueNumber.isPresent()) {
-        fieldTag.withValue(String.valueOf(this.fieldValueNumber.getAsLong()));
-      }
-    } else {
-      fieldTag.withValue(this.fieldValue);
+  }
+
+  protected LabelTag genLabelTag() {
+
+    return label()
+        .attr(Attr.FOR, this.id)
+        // If the text is screen-reader text, then we want the label to be screen-reader
+        // only.
+        .withClass(labelText.isEmpty() ? Styles.SR_ONLY : BaseStyles.INPUT_LABEL)
+        .withText(labelText.isEmpty() ? screenReaderText : labelText);
+  }
+
+  protected DivTag wrapInDivTag(Tag fieldTag, Tag labelTag, String fieldErrorsId) {
```

Comment:
I'd give this method a name more specific to what it's doing. Is this building the base container or something else? `wrapInDivTag` implies that it's a utility method, but it appears that the logic here is quite specific to the use.

Reply:
Ok, I changed to `buildBaseContainer`. Does that work?